### PR TITLE
Removing interfaces

### PR DESCRIPTION
**Pull request by sebastienros**

### CODE REVIEW

**Comment by lahma** [comment]
Does it feel faster? 😉

**Comment by sebastienros** [comment]
It doesn't at all. I will run the benchmarks when my machine is idle again, but if there is no improvement I will be torn. No interface is simpler, but the declaration of parsers is not as intuitive/readable. Though might be good still.

But it's also more explicit so less possibilities to use an untyped parser by mistake

**Comment by lahma** [comment]
The input probably needs to be big enough too so that dispatch starts to show (enough calls compared to other infra code).

File: test/Parlot.Tests/Json/JsonParser.cs [hunk 1]
```diff
@@ -7,7 +7,7 @@ namespace Parlot.Tests.Json
 {
     public class JsonParser
     {
-        private static readonly IParser<IJson> Json;
+        private static readonly Parlot.Fluent.Parser<IJson> Json;
```
Comment (lahma) [inline test/Parlot.Tests/Json/JsonParser.cs:10]: Would it make sense to call the raw version `Parser` and the fluent version `FluentParser` in main library to disambiguate the two?

**Comment by lahma** [comment]
I run performance diff between main and this branch. Seems that interface dispatch has its cost..

## Parlot.Benchmarks.ExprBench

| **Diff**|Method|Mean|Error|Gen 0|Gen 1|Gen 2|Allocated|
|------- |-------|-------:|-------|-------:|-------:|-------:|-------:|
| Old |ParlotRawSmall|741.4 ns|83.39 ns|0.1183|0.0000|0.0000|496 B|
| **New** |	| **629.9 ns (-15%)** | **49.18 ns** | **0.1183 (0%)** | **0.0000** | **0.0000** | **496 B (0%)** |
| Old |ParlotFluentSmall|1,515.8 ns|130.71 ns|0.1583|0.0000|0.0000|664 B|
| **New** |	| **1,253.2 ns (-17%)** | **185.87 ns** | **0.1602 (+1%)** | **0.0000** | **0.0000** | **672 B (+1%)** |
| Old |PidginSmall|11,479.1 ns|688.91 ns|0.1831|0.0000|0.0000|816 B|
| **New** |	| **11,292.7 ns (-2%)** | **346.06 ns** | **0.1831 (0%)** | **0.0000** | **0.0000** | **816 B (0%)** |
| Old |ParlotRawBig|3,776.2 ns|123.61 ns|0.5264|0.0000|0.0000|2208 B|
| **New** |	| **3,395.6 ns (-10%)** | **1,057.59 ns** | **0.5264 (0%)** | **0.0000** | **0.0000** | **2208 B (0%)** |
| Old |ParlotFluentBig|8,010.7 ns|347.72 ns|0.6866|0.0000|0.0000|2896 B|
| **New** |	| **6,886.6 ns (-14%)** | **651.62 ns** | **0.6943 (+1%)** | **0.0000** | **0.0000** | **2904 B (0%)** |
| Old |PidginBig|57,633.9 ns|1,462.66 ns|0.9155|0.0000|0.0000|4072 B|
| **New** |	| **54,990.8 ns (-5%)** | **4,306.38 ns** | **0.9155 (0%)** | **0.0000** | **0.0000** | **4072 B (0%)** |


## Parlot.Benchmarks.JsonBench

| **Diff**|Method|Mean|Error|Gen 0|Gen 1|Gen 2|Allocated|
|------- |-------|-------:|-------|-------:|-------:|-------:|-------:|
| Old |BigJson_Parlot|243.0 μs|26.50 μs|24.9023|0.4883|0.0000|101.8 KB|
| **New** |	| **225.3 μs (-7%)** | **12.94 μs** | **24.9023 (0%)** | **1.4648 (+200%)** | **0.0000** | **101.8 KB (0%)** |
| Old |BigJson_Pidgin|456.5 μs|7.94 μs|24.9023|3.4180|0.0000|101.7 KB|
| **New** |	| **456.4 μs (0%)** | **56.38 μs** | **24.9023 (0%)** | **3.9063 (+14%)** | **0.0000** | **101.7 KB (0%)** |
| Old |BigJson_Sprache|3,248.3 μs|272.03 μs|1308.5938|50.7813|0.0000|5349.63 KB|
| **New** |	| **3,250.1 μs (0%)** | **555.95 μs** | **1308.5938 (0%)** | **50.7813 (0%)** | **0.0000** | **5349.63 KB (0%)** |
| Old |BigJson_Superpower|1,709.1 μs|37.42 μs|222.6563|1.9531|0.0000|913.43 KB|
| **New** |	| **1,710.0 μs (0%)** | **31.14 μs** | **222.6563 (0%)** | **1.9531 (0%)** | **0.0000** | **913.43 KB (0%)** |
| Old |LongJson_Parlot|197.5 μs|4.93 μs|25.3906|0.4883|0.0000|104.34 KB|
| **New** |	| **184.7 μs (-6%)** | **3.70 μs** | **25.3906 (0%)** | **0.7324 (+50%)** | **0.0000** | **104.35 KB (0%)** |
| Old |LongJson_Pidgin|388.7 μs|4.58 μs|25.3906|2.4414|0.0000|104.25 KB|
| **New** |	| **391.1 μs (+1%)** | **3.85 μs** | **25.3906 (0%)** | **2.4414 (0%)** | **0.0000** | **104.25 KB (0%)** |
| Old |LongJson_Sprache|2,630.5 μs|82.70 μs|1054.6875|11.7188|0.0000|4311.36 KB|
| **New** |	| **2,617.3 μs (-1%)** | **14.38 μs** | **1054.6875 (0%)** | **11.7188 (0%)** | **0.0000** | **4311.36 KB (0%)** |
| Old |LongJson_Superpower|1,388.3 μs|33.08 μs|171.8750|3.9063|0.0000|706.79 KB|
| **New** |	| **1,395.9 μs (+1%)** | **66.72 μs** | **171.8750 (0%)** | **3.9063 (0%)** | **0.0000** | **706.79 KB (0%)** |
| Old |DeepJson_Parlot|162.1 μs|25.56 μs|20.0195|0.9766|0.0000|82.34 KB|
| **New** |	| **151.9 μs (-6%)** | **3.33 μs** | **20.0195 (0%)** | **0.0000 (-100%)** | **0.0000** | **82.34 KB (0%)** |
| Old |DeepJson_Pidgin|500.0 μs|10.46 μs|45.8984|0.9766|0.0000|187.79 KB|
| **New** |	| **502.3 μs (0%)** | **24.53 μs** | **45.8984 (0%)** | **0.9766 (0%)** | **0.0000** | **187.79 KB (0%)** |
| Old |DeepJson_Sprache|2,845.7 μs|354.59 μs|554.6875|218.7500|0.0000|2946.56 KB|
| **New** |	| **2,828.0 μs (-1%)** | **155.08 μs** | **562.5000 (+1%)** | **230.4688 (+5%)** | **0.0000** | **2946.56 KB (0%)** |
| Old |WideJson_Parlot|144.3 μs|7.79 μs|11.7188|0.0000|0.0000|48.52 KB|
| **New** |	| **140.8 μs (-2%)** | **5.55 μs** | **11.7188 (0%)** | **0.4883** | **0.0000** | **48.52 KB (0%)** |
| Old |WideJson_Pidgin|239.0 μs|5.40 μs|11.7188|1.2207|0.0000|48.42 KB|
| **New** |	| **241.5 μs (+1%)** | **10.15 μs** | **11.7188 (0%)** | **0.9766 (-20%)** | **0.0000** | **48.42 KB (0%)** |
| Old |WideJson_Sprache|1,493.6 μs|99.24 μs|683.5938|11.7188|0.0000|2797.28 KB|
| **New** |	| **1,524.8 μs (+2%)** | **132.08 μs** | **683.5938 (0%)** | **11.7188 (0%)** | **0.0000** | **2797.28 KB (0%)** |
| Old |WideJson_Superpower|866.0 μs|6.66 μs|112.3047|1.9531|0.0000|459.74 KB|
| **New** |	| **872.8 μs (+1%)** | **18.75 μs** | **112.3047 (0%)** | **1.9531 (0%)** | **0.0000** | **459.74 KB (0%)** |


## Parlot.Benchmarks.ParlotBenchmarks

| **Diff**|Method|Mean|Error|Gen 0|Gen 1|Gen 2|Allocated|
|------- |-------|-------:|-------|-------:|-------:|-------:|-------:|
| Old |DecodeStringWithEscapes|106.48 ns|4.345 ns|0.0286|0.0000|0.0000|120 B|
| **New** |	| **99.37 ns (-7%)** | **3.273 ns** | **0.0286 (0%)** | **0.0000** | **0.0000** | **120 B (0%)** |
| Old |DecodeStringWithoutEscapes|16.53 ns|3.916 ns|0.0000|0.0000|0.0000|0 B|
| **New** |	| **15.84 ns (-4%)** | **4.645 ns** | **0.0000** | **0.0000** | **0.0000** | **0 B** |
| Old |DecodeStringWithEscapesToString|123.21 ns|12.572 ns|0.0553|0.0000|0.0000|232 B|
| **New** |	| **121.48 ns (-1%)** | **9.800 ns** | **0.0553 (0%)** | **0.0000** | **0.0000** | **232 B (0%)** |
| Old |DecodeStringWithoutEscapesToString|32.62 ns|3.658 ns|0.0268|0.0000|0.0000|112 B|
| **New** |	| **31.30 ns (-4%)** | **6.587 ns** | **0.0268 (0%)** | **0.0000** | **0.0000** | **112 B (0%)** |

**Comment by lahma** [comment]
`ITokenResult` might be redundant too?

File: test/Parlot.Tests/Json/JsonParser.cs (hunk 1 above)
Comment (sebastienros) [reply test/Parlot.Tests/Json/JsonParser.cs:10]: I actually removed the root `Parser`, so I need to remove this prefix.

**Comment by sebastienros** [comment]
I updated the ParlotBenchmarks class to only include the parlot ones. Better to do diffs:

```
|                     Method |          Mean |         Error |       StdDev |   Gen 0 |  Gen 1 | Gen 2 | Allocated |
|--------------------------- |--------------:|--------------:|-------------:|--------:|-------:|------:|----------:|
| DecodeStringWithoutEscapes |      11.57 ns |      1.396 ns |     0.077 ns |       - |      - |     - |         - |
|    DecodeStringWithEscapes |      93.99 ns |     12.178 ns |     0.668 ns |  0.0286 |      - |     - |     120 B |
|                            |               |               |              |         |        |       |           |
|         ExpressionRawSmall |     514.04 ns |    337.304 ns |    18.489 ns |  0.1183 |      - |     - |     496 B |
|      ExpressionFluentSmall |   1,025.65 ns |    211.218 ns |    11.578 ns |  0.1583 |      - |     - |     664 B |
|           ExpressionRawBig |   2,452.62 ns |    470.934 ns |    25.813 ns |  0.5264 |      - |     - |    2208 B |
|        ExpressionFluentBig |   5,616.93 ns |    775.414 ns |    42.503 ns |  0.6866 |      - |     - |    2896 B |
|                            |               |               |              |         |        |       |           |
|                    BigJson | 196,382.67 ns | 49,986.334 ns | 2,739.920 ns | 24.9023 | 7.3242 |     - |  104240 B |
|                   DeepJson | 132,440.66 ns | 20,523.640 ns | 1,124.970 ns | 20.0195 | 0.2441 |     - |   84312 B |
|                   LongJson | 145,586.05 ns | 13,606.226 ns |   745.803 ns | 25.3906 | 6.3477 |     - |  106848 B |
|                   WideJson | 125,262.08 ns | 19,067.058 ns | 1,045.130 ns | 11.7188 | 2.1973 |     - |   49680 B |
```

```
|                     Method |          Mean |         Error |       StdDev |   Gen 0 |  Gen 1 | Gen 2 | Allocated |
|--------------------------- |--------------:|--------------:|-------------:|--------:|-------:|------:|----------:|
| DecodeStringWithoutEscapes |      11.02 ns |      2.306 ns |     0.126 ns |       - |      - |     - |         - |
|    DecodeStringWithEscapes |      96.07 ns |     55.567 ns |     3.046 ns |  0.0286 |      - |     - |     120 B |
|                            |               |               |              |         |        |       |           |
|         ExpressionRawSmall |     483.13 ns |     15.814 ns |     0.867 ns |  0.1183 |      - |     - |     496 B |
|      ExpressionFluentSmall |     906.30 ns |    125.918 ns |     6.902 ns |  0.1602 |      - |     - |     672 B |
|           ExpressionRawBig |   2,352.74 ns |    329.435 ns |    18.057 ns |  0.5264 |      - |     - |    2208 B |
|        ExpressionFluentBig |   5,523.65 ns | 19,795.510 ns | 1,085.059 ns |  0.6943 |      - |     - |    2904 B |
|                            |               |               |              |         |        |       |           |
|                    BigJson | 187,169.54 ns | 24,248.149 ns | 1,329.123 ns | 24.9023 | 7.3242 |     - |  104248 B |
|                   DeepJson | 126,975.87 ns |  8,658.275 ns |   474.589 ns | 20.0195 | 0.4883 |     - |   84320 B |
|                   LongJson | 137,165.60 ns | 15,446.368 ns |   846.668 ns | 25.3906 | 4.3945 |     - |  106856 B |
|                   WideJson | 116,276.62 ns | 16,028.836 ns |   878.595 ns | 11.8408 | 2.3193 |     - |   49688 B |
```

**Pull request merged**